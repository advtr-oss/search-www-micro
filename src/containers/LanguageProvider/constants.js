/*
 *
 * LanguageProvider constants
 *
 */

export const CHANGE_LOCALE = 'app/LanguageToggle/CHANGE_LOCALE';

export const LOCALE_KEY = 'locale'
