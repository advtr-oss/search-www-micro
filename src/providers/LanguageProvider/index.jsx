/**
 * @advtr/tidy tries to be un-opinionated, so for this we need to import `LanguageProvider` and wrap it for redux
 * */

import { connect } from 'react-redux'
import { createSelector } from 'reselect'
import { LanguageProvider } from '@advtr/tidy'

import { makeSelectLocale } from './selectors'

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale
  })
)

function mapDispatchToProps (dispatch) {
  return {
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageProvider)
