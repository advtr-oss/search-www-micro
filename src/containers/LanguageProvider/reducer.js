/*
 *
 * LanguageProvider reducer
 *
 * Now actively looks for the users language
 *
 * @note: Not the best but will work
 */

import produce from 'immer'
import get from 'lodash.get'
import split from 'lodash.split'
import includes from 'lodash.includes'

import { CHANGE_LOCALE, LOCALE_KEY } from './constants'
import { DEFAULT_LOCALE, appLocales } from '../../i18n'

// Detect user language.
const userLanguage =
  window.localStorage.getItem(LOCALE_KEY) ||
  window.navigator.language ||
  window.navigator.userLanguage

let foundLanguage = includes(appLocales, userLanguage) && userLanguage

if (!foundLanguage) {
  // Split user language in a correct format.
  const userLanguageShort = get(split(userLanguage, '-'), '0')

  // Check that the language is included in the admin configuration.
  foundLanguage = includes(appLocales, userLanguageShort) && userLanguageShort
}

export const initialState = {
  locale: foundLanguage || DEFAULT_LOCALE
}

/* eslint-disable default-case, no-param-reassign */
const languageProviderReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_LOCALE:
        window.localStorage && window.localStorage.setItem(LOCALE_KEY, action.locale)
        draft.locale = action.locale
        break
    }
  })

export default languageProviderReducer
