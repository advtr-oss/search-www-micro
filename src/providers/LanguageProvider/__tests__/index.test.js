/**
 * @jest-environment jsdom
 */

import React from 'react'
import { Provider } from 'react-redux'
import { render } from '@testing-library/react'

import LanguageProvider from '../index'

import { translationMessages } from '../../../i18n'
import configureStore from '../../../configureStore'

const initialState = {}
const store = configureStore(initialState, history)

const redux = (ui, opts) => render((
  <Provider store={store}>
    <>{ui}</>
  </Provider>
), opts)

describe('LanguageProvider', function () {
  // This mainly checks that it doesn't throw, and we get a snapshot out
  // of it, so we know it renders a child component of out choice, will
  // require a new snap each time we update translationMessages, so we know
  // they are updated too
  it('should render snapshot', function () {
    const { container } = redux(
      <LanguageProvider messages={translationMessages}>
        <div dangerouslySetInnerHTML={{ __html: JSON.stringify(translationMessages) }} />
      </LanguageProvider>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
