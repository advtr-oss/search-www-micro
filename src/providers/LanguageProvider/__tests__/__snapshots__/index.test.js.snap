// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageProvider should render snapshot 1`] = `
<div>
  {"en":{"app.containers.SearchView.placeholder":"Search the world","app.containers.SearchView.title":"Search Results","boilerplate.containers.locale.en":"English","boilerplate.containers.locale.fr":"French"},"fr":{"app.containers.SearchView.placeholder":"Cherchez le monde","app.containers.SearchView.title":"Résultats de recherche","boilerplate.containers.locale.en":"Anglaise","boilerplate.containers.locale.fr":"Français"}}
</div>
`;
